### PR TITLE
Custom header key format

### PR DESCRIPTION
**Pull request by esmet**
Support custom, case-sensitive header rewrites using case-insensitive matching

### CODE REVIEW

File: source/common/http/utility.cc [hunk 1]
```diff
@@ -420,6 +420,20 @@ Utility::parseHttp1Settings(const envoy::config::core::v3::Http1ProtocolOptions&
...
+  } else if (config.header_key_format().has_custom()) {
+    ret.header_key_format_ = Http1Settings::HeaderKeyFormat::Custom;
+    const auto &rules = config.header_key_format().custom().rules();
+    // Transform rules, inserting new elements into ret.header_key_format_rules...
+    std::transform(rules.cbegin(), rules.cend(),
+        std::inserter(ret.header_key_format_rules_, ret.header_key_format_rules_.end()),
+        [] (const auto &pair) -> std::pair<std::string, std::string> {
+          // ...by transforming each pair's key into lowercase, and using the value as is.
+          std::string lower{pair.first};
+          std::transform(lower.begin(), lower.end(), lower.begin(), [](unsigned char c) -> unsigned char {
```
Comment (esmet) [inline source/common/http/utility.cc]: TODO: Use LowerCaseString

**Review by kflynn — COMMENTED** [review]
I think I follow what's up here, and assuming that I'm correct, looks like this'll DTRT, woot! Definitely need to get together for a sync review, though. And comments. Moar comments. 😂

File: include/envoy/http/codec.h [hunk 2]
```diff
@@ -384,10 +384,14 @@ struct Http1Settings {
     // Performs proper casing of header keys: the first and all alpha characters following a
     // non-alphanumeric character is capitalized.
     ProperCase,
+    // Performs custom casing of header keys. Rules are defined
+    Custom
```
Comment (LukeShu) [inline include/envoy/http/codec.h]: Given that `ProperCase` had a trailing comma, I think style would be to include a trailing coma here.

Also, it looks like the comment is cut off?

File: source/common/http/http1/header_formatter.cc [hunk 3]
```diff
@@ -19,6 +19,21 @@ std::string ProperCaseHeaderKeyFormatter::format(absl::string_view key) const {
...
+
+std::string CustomHeaderKeyFormatter::format(absl::string_view key) const {
+  auto copy = std::string(key);
+
+  // Check for a custom header key rewrite
+  const auto &rewrite = rules_.find(copy);
+  if (rewrite != rules_.end()) {
+    // Return a copy of the rewrite.
+    return rewrite->second;
+  }
```
Comment (LukeShu) [inline source/common/http/http1/header_formatter.cc:31]: Does `.find()` return a copy? Shouldn't you be making a copy here by doing `return std::string(rewrite->second)`?

File: source/common/http/http1/codec_impl.cc [hunk 4]
```diff
@@ -59,6 +59,10 @@ HeaderKeyFormatterPtr formatter(const Http::Http1Settings& settings) {
     return std::make_unique<ProperCaseHeaderKeyFormatter>();
   }
 
+  if (settings.header_key_format_ == Http1Settings::HeaderKeyFormat::Custom) {
+    return std::make_unique<CustomHeaderKeyFormatter>(settings.header_key_format_rules_);
+  }
+
```
Comment (LukeShu) [inline source/common/http/http1/codec_impl.cc:66]: Now that there are >2 options, it probably makes sense to change this to a `switch`?

**Review by LukeShu — CHANGES_REQUESTED** [review]
Changes requested:
- cut off comment
- switch
- comma

File: source/common/http/http1/codec_impl.cc (hunk 4 above)
Comment (esmet) [reply source/common/http/http1/codec_impl.cc:66]: Done

File: include/envoy/http/codec.h (hunk 2 above)
Comment (esmet) [reply include/envoy/http/codec.h]: Done, and fixed comment

**Pull request merged**